### DISCUSSION TITLE
Access files in an AWS S3 requester pay bucket

### DISCUSSION THREAD

**Discussion by giswqs** [developmentseed/titiler#669] [Q&A]
Can Titiler access COG files in an AWS S3 requester pay bucket, such as the naip bucket?  I am able to acess the files using gdal and rasterio, but I can't find an example using TiTiler to access the tiles. 

https://registry.opendata.aws/naip/

```python
import rasterio

with rasterio.Env(AWS_REQUEST_PAYER='requester'):
    with rasterio.open('s3://naip-analytic/al/2011/100cm/rgbir_cog/30085/m_3008501_ne_16_1_20110815.tif') as src:
        print(src.profile)
```

```bash
gdalinfo /vsis3/naip-analytic/al/2011/100cm/rgbir_cog/30085/m_3008501_ne_16_1_20110815.tif \
--config AWS_REQUEST_PAYER requester
```

**Comment by vincentsarago** [comment]
@giswqs if you set `AWS_REQUEST_PAYER='requester'` in your deployment environment, titiler will be able to access requester-pays bucket 

if the question is: `can I access requester-pays bucket with titiler.xyz?` the answer is no because we can't assume the price for this

**Reply by giswqs** [reply]
Is there anything else I need to do besides adding the environment variable `AWS_REQUEST_PAYER='requester'`? I already added that and deployed the changes, but I still get the `access denied error` then `The file  does not exist in the file system, and is not recognized as a supported dataset name."
}`

![image](https://github.com/developmentseed/titiler/assets/5016453/e17c191c-7532-4a96-a05e-06ad48db6024)

![image](https://github.com/developmentseed/titiler/assets/5016453/0b38c74e-180c-4717-a683-9133c9bef32b)

**Reply by vincentsarago** [reply]
you also need to make sure the IAM role for your tile allows `naip-analytic` bucket or `*` 

https://github.com/developmentseed/titiler/blob/main/deployment/aws/cdk/app.py#L170-L176

**Reply by giswqs** [reply]
Yes, i already allows * when deloying it

![image](https://github.com/developmentseed/titiler/assets/5016453/1c924726-d835-44ba-8ca6-5b3f0fc84859)

**Reply by vincentsarago** [reply]
I just tried `AWS_REQUEST_PAYER=requester` in titiler.xyz and it worked fine 🤷‍♂️

**Reply by giswqs** [reply]
That's strange. Can you enable it temporarily and try out the s3 file below? It does not work for me. 

https://titiler.xyz/docs#/Cloud%20Optimized%20GeoTIFF/info_cog_info_get

s3://naip-analytic/al/2011/100cm/rgbir_cog/30085/m_3008501_ne_16_1_20110815.tif

**Reply by vincentsarago** [reply]
```
https://titiler.xyz/cog/info?url=s3://naip-analytic/al/2011/100cm/rgbir_cog/30085/m_3008501_ne_16_1_20110815.tif
```
<img width="824" alt="Screenshot 2023-07-18 at 3 28 27 PM" src="https://github.com/developmentseed/titiler/assets/10407788/ef060fad-e2d3-4db1-8e82-c2a9ee0e8a5d">

**Reply by giswqs** [reply]
This is the deployment settings I used. Then added `AWS_REQUEST_PAYER=requester` to the environment variables. Still getting access denied. 
![image](https://github.com/developmentseed/titiler/assets/5016453/a8577f60-34d3-4ceb-8d60-9bcf636a4d1f)

**Reply by giswqs** [reply]
I used your link. It does not work on my end. 
https://titiler.xyz/cog/info?url=s3://naip-analytic/al/2011/100cm/rgbir_cog/30085/m_3008501_ne_16_1_20110815.tif
![image](https://github.com/developmentseed/titiler/assets/5016453/1a31ce33-e1e0-4f88-ad23-9c2fd47bf22c)

**Reply by vincentsarago** [reply]
<img width="1001" alt="Screenshot 2023-07-18 at 3 34 17 PM" src="https://github.com/developmentseed/titiler/assets/10407788/edaee57b-7233-4714-a2e4-36511b7541fb">

**Reply by vincentsarago** [reply]
😬 I have no idea what's going on
<img width="983" alt="Screenshot 2023-07-18 at 3 37 59 PM" src="https://github.com/developmentseed/titiler/assets/10407788/3f851dcc-d2c0-4f6a-9aa7-26f1d941cbc6">

**Reply by giswqs** [reply]
That's puzzling. I tried the link on multiple computers and cellphone networks, none of them works. Sometimes getting this error
![image](https://github.com/developmentseed/titiler/assets/5016453/dcdd0f91-1dda-4e09-ba8f-2fe1b53c0c9b)

**Reply by vincentsarago** [reply]
Yeah I'm getting this one too ... that's new to me 😬

**Reply by vincentsarago** [reply]
Maybe it's with this dataset only 

I have no issue with `s3://naip-analytic/wa/2017/100cm/rgbir_cog/48117/m_4811762_se_11_1_20170723.tif`

**Reply by giswqs** [reply]
Something strange. The old example dataset works for me now, but the new one you just mentioned does not work for me. 
It seems if the dataset has been tried for many times, somehow it works. Newly accessed dataset does not work? 

![image](https://github.com/developmentseed/titiler/assets/5016453/ad87df65-fda2-452a-b944-b317cb823e38)


![image](https://github.com/developmentseed/titiler/assets/5016453/f423ce33-5c57-4bf0-8765-4256b8996f41)

**Comment by giswqs** [comment]
Tried it on some new datasets. Did not work either. How strange.

https://titiler.xyz/cog/info?url=s3://naip-analytic/tn/2018/60cm/rgbir_cog/34084/649969903%20m_3408401_ne_16_060_20181021.tif
![image](https://github.com/developmentseed/titiler/assets/5016453/76e98cc8-ced8-4f4a-8868-f763a96307a8)


```bash
s3 ls s3://naip-analytic/tn/2018/60cm/rgbir_cog/34084/ --request-payer 
```
![image](https://github.com/developmentseed/titiler/assets/5016453/bdeab68a-2cf3-4df8-baf7-8dce60fee7b6)

**Reply by vincentsarago** [reply]
I removed the `AWS_REQUEST_PAYER` settings from titiler.xyz, sorry

**Reply by giswqs** [reply]
I understand. I just wanted to figure out how make it work with my own deployed TiTiler, but it is not working. Sample [link](https://6ia56f0mod.execute-api.us-east-1.amazonaws.com/cog/info?url=s3://naip-analytic/al/2011/100cm/rgbir_cog/30085/m_3008501_ne_16_1_20110815.tif) to my own TiTiler.

**Reply by vincentsarago** [reply]
FYI 
you share `https://titiler.xyz/cog/info?url=s3://naip-analytic/tn/2018/60cm/rgbir_cog/34084/649969903%20m_3408401_ne_16_060_20181021.tif` but the correct URL is 
`https://titiler.xyz/cog/info?url=s3://naip-analytic/tn/2018/60cm/rgbir_cog/34084/m_3408401_ne_16_060_20181021.tif`

**Reply by giswqs** [reply]
My bad. Your titile seems to be working now. Mine is still not working. Keep trying

**Reply by solomon-negusse** [reply]
Hi @giswqs, curious if you  have been able to resolve this?  Quite puzzled by this issue as well..

**Reply by giswqs** [reply]
@solomon-negusse  See this example: https://github.com/banesullivan/localtileserver/issues/182#issuecomment-1936576886

**Reply by solomon-negusse** [reply]
Thanks, @giswqs!
